# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: widget/nsDragServiceProxy.cpp
# Commit: 2a040986845d
# Full Hash: 2a040986845da97f1eaf87c33136d9168817606a
# Author: Neil Deakin <neil@mozilla.com>
# Date: 2015-04-13 10:00:52
# Regressor Bug: 1071562
# File Overlap Count: 1
# Description:
#   Bug 1071562, e10s, support non-text types in clipboard (html, images, etc), r=smaug
# ==============================================================================

diff -r 8975ed86f399 -r 2a040986845d widget/nsDragServiceProxy.cpp
--- a/widget/nsDragServiceProxy.cpp	Fri Apr 10 09:43:38 2015 -0700
+++ b/widget/nsDragServiceProxy.cpp	Fri Apr 10 12:51:06 2015 -0400
@@ -57,21 +57,17 @@
     if (surface) {
       mozilla::RefPtr<mozilla::gfx::DataSourceSurface> dataSurface =
         surface->GetDataSurface();
-      mozilla::gfx::DataSourceSurface::MappedSurface map;
-      dataSurface->Map(mozilla::gfx::DataSourceSurface::MapType::READ, &map);
+
+      size_t length;
+      int32_t stride;
+      const uint8_t* data = nsContentUtils::GetSurfaceData(dataSurface, &length, &stride);
+      nsDependentCString dragImage(reinterpret_cast<const char*>(data), length);
+
       mozilla::gfx::IntSize size = dataSurface->GetSize();
-      mozilla::CheckedInt32 requiredBytes =
-        mozilla::CheckedInt32(map.mStride) * mozilla::CheckedInt32(size.height);
-      size_t bufLen = requiredBytes.isValid() ? requiredBytes.value() : 0;
-      mozilla::gfx::SurfaceFormat format = dataSurface->GetFormat();
-      // Surface data handling is totally nuts. This is the magic one needs to
-      // know to access the data.
-      bufLen = bufLen - map.mStride + (size.width * BytesPerPixel(format));
-      nsDependentCString dragImage(reinterpret_cast<char*>(map.mData), bufLen);
       mozilla::unused <<
         child->SendInvokeDragSession(dataTransfers, aActionType, dragImage,
-                                     size.width, size.height, map.mStride,
-                                     static_cast<uint8_t>(format),
+                                     size.width, size.height, stride,
+                                     static_cast<uint8_t>(dataSurface->GetFormat()),
                                      dragRect.x, dragRect.y);
       dataSurface->Unmap();
       StartDragSession();
